### DISCUSSION TITLE
How do I sandbox nmap?

### DISCUSSION THREAD

**Discussion by crocket** [netblue30/firejail#4980] [Q&A]
Without nonewprivs, I can slap linux capabilities to files. Thus, no root privilege is required for net_raw capability required for UDP port scanning.

With nonewprivs, net_raw capability has to come from root or setpriv. But, setpriv has difficulty with inserting capabilities through firejail, and setpriv complicates invocation. As far as I know, firejail cannot sandbox root.

How do I best sandbox nmap with firejail? I could restrict nmap with apparmor, but changing network namespace still requires sudo.
Firejail can change network namespace without sudo.

**Comment by rusty-snake** [comment]
> changing network namespace still requires sudo

Where is the difference whether you prefix sudo or firejail?

> As far as I know, firejail cannot sandbox root.

What do you mean exactly?

**Reply by crocket** [reply]
Firejail doesn't require password, and firejail simplifies invocation.
Without firejail, I need to call `sudo ip netns exec xxx sudo -u user cmd ...`.
Making specific commands passwordless sudo commands makes a mess in /etc/sudoers.d
I don't want to add hundreds of passwordless sudo commands in /etc/sudoers.d

Days ago, topimiettinen wrote

> Since file system binaries are owned by UID 0, it may be possible to overwrite /sbin/init even without any capabilities and seccomp can't help there either. Mount namespaces can be avoided by making device nodes and mounting file systems without restrictions somewhere. Still, Firejail isn't useless at all, attacks aren't always so clever and there may be enough easier targets if simple attacks fail somewhere.

> MAC systems (AppArmor, SELinux, Smack, TOMOYO) can make further restrictions on top of what Firejail provides. Rules for a daemon could only allow it to read its own configuration, write to its own logs and state files, but not for example modify state files of another daemon even if both run as UID 0, or overwrite /sbin/init even if it's owned by UID 0.

**Reply by rusty-snake** [reply]
```
# Is default IIRC
# Now calls to write() will return EROFS
read-only /usr
# Do not allow to mess things up with (u)mount
caps.drop all
nonewprivs
seccomp.drop @mount
# Hide devies nodes
private-dev
...
```

**Reply by crocket** [reply]
@rusty-snake How do you suggest I sandbox nmap with firejail? nmap requires at least net_raw for UDP port scanning.

**Reply by topimiettinen** [reply]
Firejail isn't designed at the moment to handle the case where privileges are elevated from unprivileged user level. In case of capabilities, there are five sets of capability bits: effective, inheritable, permitted, bound and ambient. Firejail's `caps.drop` makes sure that the bits are zero and `caps.keep` may retain the '1' bits, but these apply only to the effective, inheritable and permitted bits. For running nmap as unprivileged user, we'd need to raise bits in ambient set (normally all zero).

Similar use cases could be made for other privileges:
- running as some other UID than calling UID (even 0): simple app isolation from each other
- mounting file systems in a way that the user has wider access than without Firejail, for example hidden directories or file systems: more confidential information isn't accessible without Firejail and then only to specific apps
- grant network access when the original namespace didn't have any (reverse of `net none`): system setup where by default tools can't access net at all, unless via Firejail

Obviously, granting more privileges needs careful thought.

**Reply by crocket** [reply]
@topimiettinen How would you sandbox nmap with firejail if it's run as root or an unprivileged user? Is running `firejail nmap` as root secure enough?

**Reply by topimiettinen** [reply]
> @topimiettinen How would you sandbox nmap with firejail if it's run as root or an unprivileged user? Is running `firejail nmap` as root secure enough?

Then the UID is still 0, which isn't great. I don't think there is an optimal solution with Firejail (UID > 0, ambient capabilities, usual sandboxing) yet. This would be possible as systemd service (`User=`, `AmbientCapabilities=`, `InaccessiblePaths=` etc.). Also Bubblewrap has `--add-cap` and `--uid` options, don't know if they work together.

**Reply by crocket** [reply]
It seems apparmor is designed to restrict programs like nmap. If firejail has `caps.add`, then I wouldn't need root.

Without nonewprivs, file capabilities may help run nmap without root.

`caps.keep` and `caps.drop` don't make much sense because capabilities usually come from root or files.
`caps.drop all` is the only thing that makes sense for firejail.

**Reply by rusty-snake** [reply]
> If firejail has caps.add, then I wouldn't need root.

If firejail had `caps.add` it would be a security vulnerability.

**Reply by crocket** [reply]
It is difficult for firejail to contain root.
Without root or caps.add, I can drop nonewprivs, add caps.keep, and add file capabilities to /usr/bin/nmap.
I challenge you to come up with a better way.

If `caps.add` is a security vulnerability, why does bubblewrap have `--cap-add`? Perhaps, the real purpose behind bubblewrap is not security or privacy, but an electronic prison for consumers?

**Reply by rusty-snake** [reply]
> It is difficult for firejail to contain root.

difficult != impossible :wink:

So your issue is that there are (potential) ways to escape if the sandbox isn't tight enough. But this is the same case if you start it as normal user. The most sanboxes have holes. If you want a to use a sandbox without holes use `prctl(PR_SET_SECCOMP, SECCOMP_MODE_STRICT, NULL, NULL, NULL)` (and care about spectre like attacks). Accept that there will always be some weak paths but try to keep them as low as you can, then you have pretty sufficient sandbox. If you still need more security GNU/Linux and firejail are the wrong tools for you.

> If caps.add is a security vulnerability,

Depends what you exactly mean, but my understanding was that it should put capabilities in the ambient set. This is only secure if firejail is stared by root.

> why does bubblewrap have --cap-add?

`--cap-add` grands the capabilities only in the userns.

**Reply by crocket** [reply]
1. Start firejail as root
2. private-bin nmap,bash
3. private-dev,private-tmp,disable-mnt
4. private-etc aaa,bbb,ccc,...
5. caps.keep net_raw
6. Secure enough sandbox?

**Reply by topimiettinen** [reply]
> If firejail had `caps.add` it would be a security vulnerability.

Yes, if the feature would be available via command line or ~/.config, but that of course shouldn't be allowed. This feature should be available only when administrator enables it and then only for certain profiles (like nmap or ping only), probably also making the whole profile unmodifiable by the user. Then it would improve security, because nmap could be fully sandboxed, have ambient caps and still be UID > 0.

The other potential use cases could also be misused:
- UID change: steal other users' data, gain root (if UID 0 was also allowed)
- arbitrary mounting or overriding with bind-mounting for example mode 0700 directory or 0000 (no access without CAP_DAC_OVERRIDE): steal other users' data, gain root
- network access: break out of sandboxes

But with careful design, these features could be used to improve security.

> Depends what you exactly mean, but my understanding was that it should put capabilities in the ambient set. This is only secure if firejail is stared by root.

It shouldn't make too much difference whether the program is started by root (probably with sudo or from /sbin/init) or starting it as SUID. What matters is how access control is implemented and what is the end result (UID, capabilities, contents of mount and net namespaces).

**Reply by rusty-snake** [reply]
> Secure enough sandbox?

D-Bus, ...

Just use everything you can https://github.com/netblue30/firejail/blob/master/etc/templates/profile.template.

> What matters is how access control is implemented

As long as root has the control, ...

**Reply by topimiettinen** [reply]
>     6. Secure enough sandbox?

"Enough" can't be defined. This sandbox would be much better than nothing, but it would use UID 0 with limited privileges, which the attackers could use to escalate to UID 0 with full privileges (or just enough more privileges to do what they want).

The entire file hierarchy should have nothing writable to UID 0, including `/dev`, `/proc`, `/run` and `/sys`. This can be pretty hard to implement fully. If nmap doesn't need access to some of these (or their subdirectories), they should be made completely inaccessible. Seccomp should be used to the max.

**Reply by crocket** [reply]
> This feature should be available only when administrator enables it and then only for certain profiles (like nmap or ping only), probably also making the whole profile unmodifiable by the user.

This is easy. If `caps.add` was only available in /etc/firejail, not in ~/.config/firejail or command line, then it is secure.

Or, if /etc/firejail/profile.profile has `caps.allow`, profile.local or command line could specify `caps.request`.

**Reply by topimiettinen** [reply]
> > This feature should be available only when administrator enables it and then only for certain profiles (like nmap or ping only), probably also making the whole profile unmodifiable by the user.
> 
> This is easy. If `caps.add` was only available in /etc/firejail, not in ~/.config/firejail or command line, then it is secure.
> 
> Or, if /etc/firejail/profile.profile has `caps.allow`, profile.local or command line could specify `caps.request`.

That's the easy part but it's far from easy in the big picture. Current Firejail design is that the sandboxed process should always have equal or less privileges than the calling process (even if there's a SUID process in between). So it hasn't been interesting to protect the sandboxed process from the calling process (actually the opposite). The entire code base should be audited with this new goal in mind. It's been OK that user config or command line options loosen the sandbox, all that can be gained is equal privileges than there was without Firejail (assuming that Firejail can't leak extra privileges from the SUID phase). If Firejail was changed to allow more privileges to be gained, loosening the sandbox with ~/.config or command line options would be a trivial way for the attacker to gain the extra privileges for other uses besides the intended app. This may not seem very dangerous for the single user who is also admin case, but the calling process could be also subverted by an attacker, who'd want to gain more privileges.

One possible approach could be to force such profiles to use strictly only allow/whitelisting: the app should have no privileges nor access at all to file system including `/dev`, `/proc`, `/run` and `/sys`, network, seccomp, capabilities, anything, unless explicitly allowed by the config in `/etc`. Reading ~/.config and command line options should be disabled. Once the design would be understood better, perhaps some of the restrictions could lifted later. Perhaps using such a profile could be even guarded with an interactive confirmation or password prompt like UAC in Windows, with CAPTCHAs or it could require MFA. There could still be MITM-like issues to be solved. Understanding and reusing Bubblewrap's approach would be also a good start: how is CAP_NET_RAW in user namespace different from the same in global namespace?

**Reply by rusty-snake** [reply]
bubblewrap's `--cap-add` is only allowed if installed non-suid (i.e. on systems where you can simply call `unsahre(CLONE_NEWUSER)` to acquire additional capabilities in a new userns) or if it is installed suid and started by root.

https://github.com/containers/bubblewrap/blob/a362d3c0ee79097ce2e3fca4ad003f1e96dad472/bubblewrap.c#L2598-L2599

**Reply by topimiettinen** [reply]
> bubblewrap's `--cap-add` is only allowed if installed non-suid (i.e. on systems where you can simply call `unsahre(CLONE_NEWUSER)` to acquire additional capabilities in a new userns) or if it is installed suid and started by root.
> 
> https://github.com/containers/bubblewrap/blob/a362d3c0ee79097ce2e3fca4ad003f1e96dad472/bubblewrap.c#L2598-L2599

Maybe I'm missing something, but what's the value of 'real' CAP_NET_RAW if unprivileged users (no CAP_NET_RAW) can simply `unshare()` and use CAP_NET_RAW of the user namespace to capture all packets? Is that possible?

**Reply by rusty-snake** [reply]
The answer will probably something like "CAP_NET_RAW in a userns only works for in a network namespace owned by the usernamespace" but I'm not an expert here.

---

A bit OT:

Allowing sandboxed processes to `unshare(CLONE_NEWUSER)` can be danger because it gives them a full bounding set back after `--cap.drop=all`. That's why I opened #4939.

One real example would be [CVE-2022-0185](https://nvd.nist.gov/vuln/detail/CVE-2022-0185)

> Exploitation relies on the CAP_SYS_ADMIN capability; however, the permission only needs to be granted in the current namespace. An unprivileged user can use unshare(CLONE_NEWNS|CLONE_NEWUSER) to enter a namespace with the CAP_SYS_ADMIN permission, and then proceed with exploitation to root the system.

**Reply by crocket** [reply]
The easier way out for me is to sandbox root with firejail or restrict nmap with apparmor. Or, I could sandbox root with firejail and restrict nmap with apparmor simultaneously, but apparmor makes firejail largely redundant.
